File: resources/sdk/purecloudjava/tests/ApiClientRetryTest.java

package com.mypurecloud.sdk.v2;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.when;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.spy;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.common.base.Stopwatch;
import com.mypurecloud.sdk.v2.*;
import com.mypurecloud.sdk.v2.connector.*;
import com.mypurecloud.sdk.v2.extensions.AuthResponse;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.testng.annotations.Test;
import org.testng.annotations.BeforeMethod;
import org.testng.Assert;
import org.apache.http.*;
import org.apache.http.params.HttpParams;
import org.apache.http.message.BasicStatusLine;
import org.apache.http.ProtocolVersion;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpUriRequest;
import com.mypurecloud.sdk.v2.connector.apache.ApacheHttpClientConnector;

import java.util.concurrent.TimeUnit;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.HashMap;
import java.io.InputStream;
import java.io.IOException;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Type;
import java.util.Collections;

public class ApiClientRetryTest {

    private ApiClient apiClient;

    private String environment;
    private PureCloudRegionHosts region;
    private boolean useenum = true;
    Stopwatch stopwatch;

    CloseableHttpClient client;

    CloseableHttpClient spyClient;

    ApacheHttpClientConnector connector;

    @BeforeMethod
    public void setup() {
        client = HttpClientBuilder.create().build();
        spyClient = spy(client);
        MockitoAnnotations.initMocks(this);
        connector = new ApacheHttpClientConnector(spyClient, null);
    }

    @Test
    public void invokeTestWith_429() throws IOException {
        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();

        retryConfiguration.setMaxRetryTime(6);
        retryConfiguration.setDefaultDelay(100);

        apiClient = getApiClient(retryConfiguration);

        Map<String, String> headers = new HashMap<>();
        headers.put("Retry-After", "3");

        try {
            doReturn(getCloseableHttpResponse(429, headers)).when(spyClient).execute(any(HttpUriRequest.class));

            stopwatch = Stopwatch.createStarted();
            ApiResponse<ApiClientConnectorResponse> response = apiClient.invoke(getConnectorRequest(), getReturnType());
        } catch (ApiException ex) {
            Assert.assertEquals(429, ex.getStatusCode());
            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 6000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 6100, "It will wait for every 100 Mills and retry until 6 Seconds");
            stopwatch.stop();
        }
    }

    @Test
    public void invokeTestWith_429_And_No_MaxRetryTime() throws IOException {
        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        apiClient = getApiClient(retryConfiguration);

        Map<String, String> headers = new HashMap<>();
        headers.put("Retry-After", "3");

        try {
            doReturn(getCloseableHttpResponse(429, headers)).when(spyClient).execute(any(HttpUriRequest.class));

            stopwatch = Stopwatch.createStarted();
            ApiResponse<ApiClientConnectorResponse> response = apiClient.invoke(getConnectorRequest(), getReturnType());
        } catch (ApiException ex) {
            Assert.assertEquals(429, ex.getStatusCode());
            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 0 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 100, "Since maxRetryTime is not provided it will not retry even if the status code is 429");
            stopwatch.stop();
        }
    }

    @Test
    public void invokeTestWith_502() throws IOException {
        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();

        retryConfiguration.setMaxRetryTime(13);
        retryConfiguration.setDefaultDelay(2000);
        retryConfiguration.setBackoffInterval(11000);

        apiClient = getApiClient(retryConfiguration);

        try {
            doReturn(getCloseableHttpResponse(502, Collections.emptyMap())).when(spyClient).execute(any(HttpUriRequest.class));

            stopwatch = Stopwatch.createStarted();
            ApiResponse<ApiClientConnectorResponse> response = apiClient.invoke(getConnectorRequest(), getReturnType());
        } catch (ApiException ex) {
            Assert.assertEquals(502, ex.getStatusCode());
            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 13000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 13100L, "It will wait for every 2 Sec and retry for 5 times then it will backoff for 3 sec and retry then it exits.");
            stopwatch.stop();
        }
    }

    @Test
    public void invokeTestWith_503() throws IOException {
        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();

        retryConfiguration.setMaxRetryTime(40);
        retryConfiguration.setDefaultDelay(200);
        retryConfiguration.setBackoffInterval(3000);

        apiClient = getApiClient(retryConfiguration);

        try {
            doReturn(getCloseableHttpResponse(503, Collections.emptyMap())).when(spyClient).execute(any(HttpUriRequest.class));

            stopwatch = Stopwatch.createStarted();
            ApiResponse<ApiClientConnectorResponse> response = apiClient.invoke(getConnectorRequest(), getReturnType());
        } catch (ApiException ex) {
            Assert.assertEquals(503, ex.getStatusCode());
            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 40000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 40100, "It will wait for every 200 Mills and retry for 5 times then it will backoff for 3 Sec once, 9 Sec once and 27 Sec before retrying");
            stopwatch.stop();
        }
    }


    @Test
    public void invokeTestWith_504() throws IOException {
        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();

        retryConfiguration.setMaxRetryTime(2);
        retryConfiguration.setDefaultDelay(1000);

        apiClient = getApiClient(retryConfiguration);

        try {
            doReturn(getCloseableHttpResponse(504, Collections.emptyMap())).when(spyClient).execute(any(HttpUriRequest.class));

            stopwatch = Stopwatch.createStarted();
            ApiResponse<ApiClientConnectorResponse> response = apiClient.invoke(getConnectorRequest(), getReturnType());
        } catch (ApiException ex) {
            Assert.assertEquals(504, ex.getStatusCode());
            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 2000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2100, "It will wait for every 1 sec and retry for 2 times");
            stopwatch.stop();
        }
    }

    @Test
    public void invokeTestWith_504_No_MaxRetryTime() throws IOException {
        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();

        apiClient = getApiClient(retryConfiguration);

        try {
            doReturn(getCloseableHttpResponse(504, Collections.emptyMap())).when(spyClient).execute(any(HttpUriRequest.class));

            stopwatch = Stopwatch.createStarted();
            ApiResponse<ApiClientConnectorResponse> response = apiClient.invoke(getConnectorRequest(), getReturnType());
        } catch (ApiException ex) {
            Assert.assertEquals(504, ex.getStatusCode());
            Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 0 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 100, "Since maxRetryTime is not provided it will not retry even if the status code is 504");
            stopwatch.stop();
        }
    }

    @Test
    public void shouldRetryTestWith_200() throws IOException {

        ApiClientConnectorResponse response = getConnectorResponse(200, Collections.emptyMap());

        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        retryConfiguration.setMaxRetryTime(30);

        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
        Stopwatch stopwatch = Stopwatch.createStarted();
        boolean result = retry.shouldRetry(response);
        Assert.assertFalse(result, "Status Code is 200, so shouldRetry method returns false");
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 0 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 100);
        stopwatch.stop();
    }

    @Test
    public void shouldRetryTestWith_429() throws IOException {
        Map<String, String> headers = new HashMap<>();
        headers.put("Retry-After", "1");

        ApiClientConnectorResponse response = getConnectorResponse(429, headers);

        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        retryConfiguration.setMaxRetryTime(30);

        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
        Stopwatch stopwatch = Stopwatch.createStarted();
        boolean result = retry.shouldRetry(response);
        Assert.assertTrue(result, "Status Code is 429, so it will sleep for 1 Sec as provided in Retry-After header and returns true");
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 1000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 1100);
        stopwatch.stop();

    }

    @Test
    public void shouldRetryTestWith_502() throws IOException {

        Map<String, String> headers = new HashMap<>();
        headers.put("Retry-After", "1");

        ApiClientConnectorResponse response = getConnectorResponse(502, headers);

        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        retryConfiguration.setMaxRetryTime(30);

        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
        Stopwatch stopwatch = Stopwatch.createStarted();
        boolean result = retry.shouldRetry(response);
        Assert.assertTrue(result, "Status Code is 502, so it will sleep for 1 Sec as provided in Retry-After header and returns true");
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 1000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 1100);
        stopwatch.stop();
    }

    @Test
    public void shouldRetryTestWith_502_And_0_MaxRetryTime() throws IOException {

        ApiClientConnectorResponse response = getConnectorResponse(502, Collections.emptyMap());

        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        Stopwatch stopwatch = Stopwatch.createStarted();
        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
        boolean result = retry.shouldRetry(response);
        Assert.assertFalse(result, "Even though Status Code is 502, it will return false because MaxRetryTime is set to Zero by default");
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 0 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 100);
        stopwatch.stop();
    }

    @Test
    public void shouldRetryTestWith_503_And_RetryConfig() throws IOException {

        ApiClientConnectorResponse response = getConnectorResponse(503, Collections.emptyMap());

        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        retryConfiguration.setBackoffInterval(6000L);
        retryConfiguration.setDefaultDelay(3000L);
        retryConfiguration.setMaxRetryTime(10);
        retryConfiguration.setMaxRetriesBeforeBackoff(0);

        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
        Stopwatch stopwatch = Stopwatch.createStarted();
        boolean result = retry.shouldRetry(response);
        Assert.assertTrue(result, "Since Status Code is 503 and maxRetriesBeforeBackoff is Zero, backoff block will be executed and returns true");
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 3000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 3100);
        stopwatch.stop();
    }

    @Test
    public void shouldRetryTestWith_504_And_No_RetryAfter_Header() throws IOException {

        ApiClientConnectorResponse response = getConnectorResponse(504, Collections.emptyMap());

        ApiClient.RetryConfiguration retryConfiguration = new ApiClient.RetryConfiguration();
        retryConfiguration.setMaxRetryTime(30);
        ApiClient.Retry retry = new ApiClient.Retry(retryConfiguration);
        Stopwatch stopwatch = Stopwatch.createStarted();
        boolean result = retry.shouldRetry(response);
        Assert.assertTrue(result, "Even though Retry-After header is missing, it will sleep for 3 Sec by default and returns true");
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) >= 3000 && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 3100);
        stopwatch.stop();
    }

    private ApiClient getApiClient(ApiClient.RetryConfiguration retryConfiguration) {
        region = getEnvironment();
        if (region == null) {
            useenum = false;
        }
        try {
            ApiClient.Builder builder = ApiClient.Builder.standard();
            builder = builder.withProperty(ApiClientConnectorProperty.CONNECTOR_PROVIDER, connector);
            if (useenum) {
                builder = builder.withBasePath(region);
            } else {
                builder = builder.withBasePath("https://api." + environment);
            }
            builder = builder.withRetryConfiguration(retryConfiguration);
            apiClient = builder.build();

            ApiResponse<AuthResponse> authResponse = apiClient.authorizeClientCredentials(getClientId(), getClientSecret());

            Configuration.setDefaultApiClient(apiClient);

        } catch (ApiException ex) {
            handleApiException(ex);
        } catch (Exception ex) {
            System.out.println(ex);
            Assert.fail();
        }
        return apiClient;

    }

    private CloseableHttpResponse getCloseableHttpResponse(int statusCode, Map<String, String> headerMap) {
        return new CloseableHttpResponse() {
            @Override
            public void close() throws IOException {

            }

            @Override
            public StatusLine getStatusLine() {
                return new BasicStatusLine(new ProtocolVersion("HTTP/1.1", 1, 1), statusCode, "Bad GateWay");
            }

            @Override
            public void setStatusLine(StatusLine statusline) {
            }

            @Override
            public void setStatusLine(ProtocolVersion ver, int code) {
            }

            @Override
            public void setStatusLine(ProtocolVersion ver, int code, String reason) {

            }

            @Override
            public void setStatusCode(int code) throws IllegalStateException {
            }

            @Override
            public void setReasonPhrase(String reason) throws IllegalStateException {

            }

            @Override
            public HttpEntity getEntity() {
                return null;
            }

            @Override
            public void setEntity(HttpEntity entity) {

            }

            @Override
            public Locale getLocale() {
                return null;
            }

            @Override
            public void setLocale(Locale loc) {

            }

            @Override
            public ProtocolVersion getProtocolVersion() {
                return null;
            }

            @Override
            public boolean containsHeader(String name) {
                return false;
            }

            @Override
            public Header[] getHeaders(String name) {
                return new Header[0];
            }

            @Override
            public Header getFirstHeader(String name) {
                return null;
            }

            @Override
            public Header getLastHeader(String name) {
                return null;
            }

            @Override
            public Header[] getAllHeaders() {
                Header header = new Header() {
                    @Override
                    public String getName() {
                        String name = null;
                        if (!headerMap.isEmpty()) {
                            for (String key : headerMap.keySet()) {
                                name = key;
                            }
                        }
                        return name;
                    }

                    @Override
                    public String getValue() {
                        String value = null;
                        if (!headerMap.isEmpty()) {
                            value = headerMap.get("Retry-After");
                        }
                        return value;
                    }

                    @Override
                    public HeaderElement[] getElements() throws ParseException {
                        return new HeaderElement[0];
                    }
                };
                Header[] allHeader = new Header[1];
                allHeader[0] = header;
                return allHeader;
            }

            @Override
            public void addHeader(Header header) {

            }

            @Override
            public void addHeader(String name, String value) {

            }

            @Override
            public void setHeader(Header header) {

            }

            @Override
            public void setHeader(String name, String value) {

            }

            @Override
            public void setHeaders(Header[] headers) {

            }

            @Override
            public void removeHeader(Header header) {

            }

            @Override
            public void removeHeaders(String name) {

            }

            @Override
            public HeaderIterator headerIterator() {
                return null;
            }

            @Override
            public HeaderIterator headerIterator(String name) {
                return null;
            }

            @Override
            public HttpParams getParams() {
                return null;
            }

            @Override
            public void setParams(HttpParams params) {

            }
        };
    }

    private ApiClientConnectorResponse getConnectorResponse(int statusCode, Map<String, String> headers) {
        return new ApiClientConnectorResponse() {
            @Override
            public int getStatusCode() {
                return statusCode;
            }

            @Override
            public String getStatusReasonPhrase() {
                return null;
            }

            @Override
            public Map<String, String> getHeaders() {
                return headers;
            }

            @Override
            public boolean hasBody() {
                return false;
            }

            @Override
            public String readBody() throws IOException {
                return null;
            }

            @Override
            public InputStream getBody() throws IOException {
                return null;
            }

            @Override
            public void close() throws Exception {

            }
        };
    }

    private ApiRequest<?> getConnectorRequest() {
        return new ApiRequest<ApiClientConnectorRequest>() {
            @Override
            public String getPath() {
                return "/api/v2/users" ;
            }

            @Override
            public String getMethod() {
                return "GET" ;
            }

            @Override
            public Map<String, String> getPathParams() {
                return Collections.emptyMap();
            }

            @Override
            public List<Pair> getQueryParams() {
                return Collections.emptyList();
            }

            @Override
            public Map<String, Object> getFormParams() {
                return Collections.emptyMap();
            }

            @Override
            public Map<String, String> getHeaderParams() {
                return Collections.emptyMap();
            }

            @Override
            public Map<String, String> getCustomHeaders() {
                return Collections.emptyMap();
            }

            @Override
            public String getContentType() {
                return "application/json" ;
            }

            @Override
            public String getAccepts() {
                return "application/json" ;
            }

            @Override
            public ApiClientConnectorRequest getBody() {
                return null;
            }

            @Override
            public String[] getAuthNames() {
                return new String[]{"PureCloud OAuth"};
            }
        };
    }

    private TypeReference<ApiClientConnectorResponse> getReturnType() {
        return new TypeReference<ApiClientConnectorResponse>() {
            @Override
            public Type getType() {
                return super.getType();
            }
        };
    }

    private PureCloudRegionHosts getEnvironment() {
        environment = System.getenv("PURECLOUD_ENVIRONMENT");
        switch (environment) {
            case "mypurecloud.com":
                return PureCloudRegionHosts.us_east_1;
            case "mypurecloud.ie":
                return PureCloudRegionHosts.eu_west_1;
            case "mypurecloud.com.au":
                return PureCloudRegionHosts.ap_southeast_2;
            case "mypurecloud.jp":
                return PureCloudRegionHosts.ap_northeast_1;
            case "mypurecloud.de":
                return PureCloudRegionHosts.eu_central_1;
            default:
                System.out.println("Not in PureCloudRegionHosts using string value");
                return null;
        }
    }

    private String getClientId() {
        return System.getenv("PURECLOUD_CLIENT_ID");
    }

    private String getClientSecret() {
        return System.getenv("PURECLOUD_CLIENT_SECRET");
    }

    private void handleApiException(ApiException ex) {
        System.out.println(ConsoleColors.applyTag(ConsoleColors.RED_BOLD, "API Exception") +
                "(" + ex.getCorrelationId() + ") " +
                ex.getStatusCode() + " " + ex.getStatusReasonPhrase() + " - " + ex.getRawBody());
        Assert.fail(ex.getStatusCode() + " " + ex.getStatusReasonPhrase() + " (" + ex.getCorrelationId() + ")");
    }
}

File: resources/sdk/purecloudkotlin/tests/ApiClientRetryTest.kt
package com.mypurecloud.sdk.v2

import com.fasterxml.jackson.core.type.TypeReference
import com.google.common.base.Stopwatch
import com.mypurecloud.sdk.v2.ApiClient.RetryConfiguration
import com.mypurecloud.sdk.v2.ConsoleColors.applyTag
import com.mypurecloud.sdk.v2.connector.ApiClientConnectorProperty
import com.mypurecloud.sdk.v2.connector.ApiClientConnectorRequest
import com.mypurecloud.sdk.v2.connector.ApiClientConnectorResponse
import com.mypurecloud.sdk.v2.connector.apache.ApacheHttpClientConnector
import org.apache.http.Header
import org.apache.http.ProtocolVersion
import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpUriRequest
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClientBuilder
import org.apache.http.message.BasicStatusLine
import org.mockito.Matchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import org.testng.Assert
import org.testng.annotations.BeforeMethod
import org.testng.annotations.Test
import java.io.IOException
import java.io.InputStream
import java.util.*
import java.util.concurrent.TimeUnit

class ApiClientRetryTest {
    private var apiClient: ApiClient? = null
    private var environment: String? = null
    private var region: PureCloudRegionHosts? = null
    private var useenum: Boolean = true
    var stopwatch: Stopwatch? = null
    var client: CloseableHttpClient? = null
    var spyClient: CloseableHttpClient? = null
    var connector: ApacheHttpClientConnector? = null
    @Mock
    private var mockResponse: CloseableHttpResponse? = null

    @BeforeMethod
    fun setup() {
        client = HttpClientBuilder.create().build()
        spyClient = Mockito.spy(client)
        MockitoAnnotations.initMocks(this)
        connector = ApacheHttpClientConnector((spyClient)!!, null)
    }

    @Test
    @Throws(IOException::class)
    fun invokeTestWith_429() {
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 6
        retryConfiguration.retryAfterDefaultMs = 100
        apiClient = getApiClient(retryConfiguration)
        mockResponse = getMockCloseableHttpResponse(429)
        val header: Header = Mockito.mock(Header::class.java)
        Mockito.`when`(header.name).thenReturn("Retry-After")
        Mockito.`when`(header.value).thenReturn("3")
        Mockito.`when`(mockResponse!!.allHeaders).thenReturn(arrayOf(header))
        Mockito.doReturn(mockResponse).`when`(spyClient)!!.execute(Matchers.any(HttpUriRequest::class.java))
        try {
            stopwatch = Stopwatch.createStarted()
            apiClient!!.invoke(connectorRequest, returnType)
        } catch (ex: ApiException) {
            Assert.assertEquals(429, ex.statusCode)
            Assert.assertTrue(stopwatch!!.elapsed(TimeUnit.MILLISECONDS) in 6000..6099, "It will wait for every 100 Mills and retry until 6 Seconds")
            stopwatch!!.stop()
        }
    }

    @Test
    @Throws(IOException::class)
    fun invokeTestWith_429_And_No_MaxRetryTime() {
        val retryConfiguration = RetryConfiguration()
        apiClient = getApiClient(retryConfiguration)
        mockResponse = getMockCloseableHttpResponse(429)
        val header: Header = Mockito.mock(Header::class.java)
        Mockito.`when`(header.name).thenReturn("Retry-After")
        Mockito.`when`(header.value).thenReturn("3")
        Mockito.`when`(mockResponse!!.allHeaders).thenReturn(arrayOf(header))
        Mockito.doReturn(mockResponse).`when`(spyClient)!!.execute(Matchers.any(HttpUriRequest::class.java))
        try {
            stopwatch = Stopwatch.createStarted()
            apiClient!!.invoke(connectorRequest, returnType)
        } catch (ex: ApiException) {
            Assert.assertEquals(429, ex.statusCode)
            Assert.assertTrue(stopwatch!!.elapsed(TimeUnit.MILLISECONDS) in 0..99, "Since maxRetryTime is not provided it will not retry even if the status code is 429")
            stopwatch!!.stop()
        }
    }

    @Test
    @Throws(IOException::class)
    fun invokeTestWith_502() {
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 13
        retryConfiguration.retryAfterDefaultMs = 2000
        retryConfiguration.backoffIntervalMs = 11000
        apiClient = getApiClient(retryConfiguration)
        mockResponse = getMockCloseableHttpResponse(502)
        Mockito.`when`(mockResponse!!.allHeaders).thenReturn(arrayOf())
        Mockito.doReturn(mockResponse).`when`(spyClient)!!.execute(Matchers.any(HttpUriRequest::class.java))
        try {
            stopwatch = Stopwatch.createStarted()
            apiClient!!.invoke(connectorRequest, returnType)
        } catch (ex: ApiException) {
            Assert.assertEquals(502, ex.statusCode)
            Assert.assertTrue(stopwatch!!.elapsed(TimeUnit.MILLISECONDS) in 13000..13099, "It will wait for every 2 Sec and retry for 5 times then it will backoff for 3 sec and retry then it exits.")
            stopwatch!!.stop()
        }
    }

    @Test
    @Throws(IOException::class)
    fun invokeTestWith_503() {
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 40
        retryConfiguration.retryAfterDefaultMs = 200
        retryConfiguration.backoffIntervalMs = 3000
        apiClient = getApiClient(retryConfiguration)
        mockResponse = getMockCloseableHttpResponse(503)
        Mockito.`when`(mockResponse!!.allHeaders).thenReturn(arrayOf())
        Mockito.doReturn(mockResponse).`when`(spyClient)!!.execute(Matchers.any(HttpUriRequest::class.java))
        try {
            stopwatch = Stopwatch.createStarted()
            apiClient!!.invoke(connectorRequest, returnType)
        } catch (ex: ApiException) {
            Assert.assertEquals(503, ex.statusCode)
            Assert.assertTrue(stopwatch!!.elapsed(TimeUnit.MILLISECONDS) in 40000..40099, "It will wait for every 200 Mills and retry for 5 times then it will backoff for 3 Sec once, 9 Sec once and 27 Sec before retrying")
            stopwatch!!.stop()
        }
    }

    @Test
    @Throws(IOException::class)
    fun invokeTestWith_504() {
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 2
        retryConfiguration.retryAfterDefaultMs = 1000
        apiClient = getApiClient(retryConfiguration)
        mockResponse = getMockCloseableHttpResponse(504)
        Mockito.`when`(mockResponse!!.allHeaders).thenReturn(arrayOf())
        Mockito.doReturn(mockResponse).`when`(spyClient)!!.execute(Matchers.any(HttpUriRequest::class.java))
        try {
            stopwatch = Stopwatch.createStarted()
            apiClient!!.invoke(connectorRequest, returnType)
        } catch (ex: ApiException) {
            Assert.assertEquals(504, ex.statusCode)
            Assert.assertTrue(stopwatch!!.elapsed(TimeUnit.MILLISECONDS) in 2000..2099, "It will wait for every 1 sec and retry for 2 times")
            stopwatch!!.stop()
        }
    }

    @Test
    @Throws(IOException::class)
    fun invokeTestWith_504_No_MaxRetryTime() {
        val retryConfiguration = RetryConfiguration()
        apiClient = getApiClient(retryConfiguration)
        mockResponse = getMockCloseableHttpResponse(504)
        Mockito.`when`(mockResponse!!.allHeaders).thenReturn(arrayOf())
        Mockito.doReturn(mockResponse).`when`(spyClient)!!.execute(Matchers.any(HttpUriRequest::class.java))
        try {
            stopwatch = Stopwatch.createStarted()
            apiClient!!.invoke(connectorRequest, returnType)
        } catch (ex: ApiException) {
            Assert.assertEquals(504, ex.statusCode)
            Assert.assertTrue(stopwatch!!.elapsed(TimeUnit.MILLISECONDS) in 0..99, "Since maxRetryTime is not provided it will not retry even if the status code is 504")
            stopwatch!!.stop()
        }
    }

    @Test
    @Throws(IOException::class)
    fun shouldRetryTestWith_200() {
        val response: ApiClientConnectorResponse = getConnectorResponse(200, emptyMap())
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 30
        val retry: ApiClient.Retry = ApiClient.Retry(retryConfiguration)
        val stopwatch: Stopwatch = Stopwatch.createStarted()
        val result: Boolean = retry.shouldRetry(response)
        Assert.assertFalse(result, "Status Code is 200, so shouldRetry method returns false")
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) in 0..99)
        stopwatch.stop()
    }

    @Test
    @Throws(IOException::class)
    fun shouldRetryTestWith_429() {
        val headers: MutableMap<String, String> = HashMap()
        headers["Retry-After"] = "1"
        val response: ApiClientConnectorResponse = getConnectorResponse(429, headers)
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 30
        val retry: ApiClient.Retry = ApiClient.Retry(retryConfiguration)
        val stopwatch: Stopwatch = Stopwatch.createStarted()
        val result: Boolean = retry.shouldRetry(response)
        Assert.assertTrue(result, "Status Code is 429, so it will sleep for 1 Sec as provided in Retry-After header and returns true")
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) in 1000..1099)
        stopwatch.stop()
    }

    @Test
    @Throws(IOException::class)
    fun shouldRetryTestWith_502() {
        val headers: MutableMap<String, String> = HashMap()
        headers["Retry-After"] = "1"
        val response: ApiClientConnectorResponse = getConnectorResponse(502, headers)
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 30
        val retry: ApiClient.Retry = ApiClient.Retry(retryConfiguration)
        val stopwatch: Stopwatch = Stopwatch.createStarted()
        val result: Boolean = retry.shouldRetry(response)
        Assert.assertTrue(result, "Status Code is 502, so it will sleep for 1 Sec as provided in Retry-After header and returns true")
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) in 1000..1099)
        stopwatch.stop()
    }

    @Test
    @Throws(IOException::class)
    fun shouldRetryTestWith_502_And_0_MaxRetryTime() {
        val response: ApiClientConnectorResponse = getConnectorResponse(502, emptyMap())
        val retryConfiguration = RetryConfiguration()
        val stopwatch: Stopwatch = Stopwatch.createStarted()
        val retry: ApiClient.Retry = ApiClient.Retry(retryConfiguration)
        val result: Boolean = retry.shouldRetry(response)
        Assert.assertFalse(result, "Even though Status Code is 502, it will return false because MaxRetryTime is set to Zero by default")
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) in 0..99)
        stopwatch.stop()
    }

    @Test
    @Throws(IOException::class)
    fun shouldRetryTestWith_503_And_RetryConfig() {
        val response: ApiClientConnectorResponse = getConnectorResponse(503, emptyMap())
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.backoffIntervalMs = 6000L
        retryConfiguration.retryAfterDefaultMs = 3000L
        retryConfiguration.maxRetryTimeSec = 10
        retryConfiguration.maxRetriesBeforeBackoff = 0
        val retry: ApiClient.Retry = ApiClient.Retry(retryConfiguration)
        val stopwatch: Stopwatch = Stopwatch.createStarted()
        val result: Boolean = retry.shouldRetry(response)
        Assert.assertTrue(result, "Since Status Code is 503 and maxRetriesBeforeBackoff is Zero, backoff block will be executed and returns true")
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) in 3000..3099)
        stopwatch.stop()
    }

    @Test
    @Throws(IOException::class)
    fun shouldRetryTestWith_504_And_No_RetryAfter_Header() {
        val response: ApiClientConnectorResponse = getConnectorResponse(504, emptyMap())
        val retryConfiguration = RetryConfiguration()
        retryConfiguration.maxRetryTimeSec = 30
        val retry: ApiClient.Retry = ApiClient.Retry(retryConfiguration)
        val stopwatch: Stopwatch = Stopwatch.createStarted()
        val result: Boolean = retry.shouldRetry(response)
        Assert.assertTrue(result, "Even though Retry-After header is missing, it will sleep for 3 Sec by default and returns true")
        Assert.assertTrue(stopwatch.elapsed(TimeUnit.MILLISECONDS) in 3000..3099)
        stopwatch.stop()
    }

    private fun getApiClient(retryConfiguration: RetryConfiguration): ApiClient? {
        region = getEnvironment()
        if (region == null) {
            useenum = false
        }
        try {
            var builder: ApiClient.Builder = ApiClient.Builder.standard()
            builder = builder.withProperty(ApiClientConnectorProperty.CONNECTOR_PROVIDER, connector)
            builder = if (useenum) {
                builder.withBasePath((region)!!)
            } else {
                builder.withBasePath("https://api.$environment")
            }
            builder = builder.withRetryConfiguration(retryConfiguration)
            apiClient = builder.build()
            apiClient!!.authorizeClientCredentials(clientId, clientSecret)
            Configuration.defaultApiClient = apiClient
        } catch (ex: ApiException) {
            handleApiException(ex)
        } catch (ex: Exception) {
            println(ex)
            Assert.fail()
        }
        return apiClient
    }

    private fun getMockCloseableHttpResponse(statusCode: Int): CloseableHttpResponse {
        val mockCloseableHttpResponse: CloseableHttpResponse = Mockito.mock(CloseableHttpResponse::class.java)
        Mockito.`when`(mockCloseableHttpResponse.statusLine).thenReturn(BasicStatusLine(ProtocolVersion("HTTP/1.1", 1, 1), statusCode, "Not Okay"))
        return mockCloseableHttpResponse
    }

    private fun getConnectorResponse(statusCode: Int, headers: Map<String, String>): ApiClientConnectorResponse {
        return object : ApiClientConnectorResponse {
            override val statusCode: Int
                get() = statusCode

            override val statusReasonPhrase: String?
                get() = null

            override fun getHeaders(): Map<String, String> {
                return headers
            }

            override fun hasBody(): Boolean {
                return false
            }

            @Throws(IOException::class)
            override fun readBody(): String? {
                return null
            }

            @Throws(IOException::class)
            override fun getBody(): InputStream? {
                return null
            }

            @Throws(Exception::class)
            override fun close() {
            }
        }
    }

    private val connectorRequest: ApiRequest<*>
        get() = object : ApiRequest<ApiClientConnectorRequest?> {
            override val path: String
                get() = "/api/v2/users"

            override val method: String
                get() = "GET"

            override val pathParams: Map<String, String>
                get() = emptyMap()

            override val queryParams: List<Pair>
                get() = emptyList()

            override val formParams: Map<String, Any>
                get() = emptyMap()

            override val headerParams: Map<String, String>
                get() = emptyMap()

            override val customHeaders: Map<String, String>
                get() = emptyMap()

            override val contentType: String
                get() = "application/json"

            override val accepts: String
                get() = "application/json"

            override val body: ApiClientConnectorRequest?
                get() = null

            override val authNames: Array<String?>
                get() = arrayOf("PureCloud OAuth")
        }

    private val returnType: TypeReference<ApiClientConnectorResponse>
        get() {
            return object : TypeReference<ApiClientConnectorResponse>() {}
        }

    private fun getEnvironment(): PureCloudRegionHosts? {
        environment = System.getenv("PURECLOUD_ENVIRONMENT")
        return when (environment) {
            "mypurecloud.com" -> PureCloudRegionHosts.us_east_1
            "mypurecloud.ie" -> PureCloudRegionHosts.eu_west_1
            "mypurecloud.com.au" -> PureCloudRegionHosts.ap_southeast_2
            "mypurecloud.jp" -> PureCloudRegionHosts.ap_northeast_1
            "mypurecloud.de" -> PureCloudRegionHosts.eu_central_1
            else -> {
                println("Not in PureCloudRegionHosts using string value")
                null
            }
        }
    }

    private val clientId: String
        get() {
            return System.getenv("PURECLOUD_CLIENT_ID")
        }

    private val clientSecret: String
        get() {
            return System.getenv("PURECLOUD_CLIENT_SECRET")
        }

    private fun handleApiException(ex: ApiException) {
        println((applyTag(ConsoleColors.RED_BOLD, "API Exception") +
                "(" + ex.getCorrelationId() + ") " +
                ex.statusCode + " " + ex.getStatusReasonPhrase() + " - " + ex.rawBody))
        Assert.fail(ex.statusCode.toString() + " " + ex.getStatusReasonPhrase() + " (" + ex.getCorrelationId() + ")")
    }
}
